qlib/FsUtil.qm

```diff
@@ -405,6 +405,53 @@ public namespace FsUtil {
         throw "FILE-WRITE-ERROR", "Impossible to find a unique name.";
     }
 
+    #! Class implementing a user friendly temporary directory creation - users don't need to care about removing it
+    public class TmpDir {
+
+        public {
+            string path;
+        }
+
+        #! Creates a unique temporary directory and returns its absolute path
+        /** @param prefix prefix to be used in the newly created directory name
+            @param suffix suffix to be used in the newly created directory name
+            @param path path to a dir in which the caller wants to create the new directory; if not provided, standard temp
+                        location will be used
+        */
+        public constructor(*string prefix, *string suffix, *string path) {
+            self.path = make_tmp_dir(prefix, suffix, path);
+        }
+
+        public destructor() {
+            remove_tree(path);
+        }
+    }
+
+    #! Class implementing a user friendly temporary file creation - users don't need to care about removing it
+    public class TmpFile {
+
+        public {
+            string path;
+            File file;
+        }
+
+        #! Creates and opens a unique temporary file and returns its absolute path as well as its File object
+        /** @param prefix prefix to be used in the newly created file name
+            @param suffix suffix to be used in the newly created file name
+            @param path path to a dir in which the caller wants to create the new file; if not provided, standard temp
+                        location will be used
+        */
+        public constructor(*string prefix, *string suffix, *string path) {
+            hash f = make_tmp_file(prefix, suffix, path);
```

Comment:
using `hash<auto>` is faster then only `hash` (why and how ask David, something with `no type stripping takes place` etc, so it's probably a good reason to use it (also on FSUtils.qm probably, I didn't notice it there, sorry)

Reply:
declaring an lvalue as `hash` or `list` still strip any complex types from any value passed to it at runtime for backwards compatibility.  This was a compromise implemented to allow us to introduce complex types everywhere in Qore.  If we don't strip the complex types with these declarations, then we get breakage which would have limited the usefulness of complex types.  Therefore `hash<auto>` is preferred over `hash` and `list<auto>` and `softlist<auto>` are preferred over `list` and `softlist`.  I'll add an entry to the wiki about this.

Reply:
wiki entry added: https://github.com/qorelanguage/qore/wiki/ProTips#use-of-complex-types